scalyr_agent/profiler.py

```diff
@@ -106,86 +236,215 @@ def _start(self, config, current_time):
+    """
+    Thread which periodically captures memory summary using pympler package.
+    """
+
+    def __init__(self, capture_interval=10, *args, **kwargs):
+        # type: (int, Any, Any) -> None
         """
-            Updates the state of the profiler - either enabling or disabling it, based on
-            the current time and whether or not the current profiling interval has started/stopped
+        :param capture_interval: How often to capture memory usage snapshot.
+        :type capture_interval: ``int``
         """
-        # no profiling if the profiler isn't available
-        if yappi is None:
-            return
+        super(PeriodicMemorySummaryCaptureThread, self).__init__(*args, **kwargs)
 
-        if current_time is None:
-            current_time = time.time()
+        self._capture_interval = capture_interval
 
-        try:
-            # check if profiling is enabled in the config and turn it on/off if necessary
-            if config.enable_profiling:
-                if not self._is_enabled:
-                    self._update_start_interval(config, current_time)
-                    self._is_enabled = True
-            else:
-                if yappi.is_running():
-                    self._stop(config, current_time)
-                self._is_enabled = False
+        self._running = False
+        self._sleep_interval = 5
+        self._profiling_data = []  # type: List[Dict[str, Any]]
 
-            # only do profiling if we are still enabled
-            if not self._is_enabled:
-                return
+        self._tracker = tracker.SummaryTracker()
 
-            # check if the current profiling interval needs to start or stop
-            if yappi.is_running():
-                if current_time > self._profile_end:
-                    self._stop(config, current_time)
-                    self._update_start_interval(config, current_time)
+    def run(self):
+        # type: () -> None
+        self._running = True
 
-            else:
-                if current_time > self._profile_start:
-                    self._start(config, current_time)
-        except Exception as e:
-            global_log.log(
-                scalyr_logging.DEBUG_LEVEL_0,
-                "Failed to update profiler: %s, %s"
-                % (six.text_type(e), traceback.format_exc()),
-                limit_once_per_x_secs=300,
-                limit_key="profiler-update",
+        last_capture_time = 0
+        while self._running:
+            if (time.time() - self._capture_interval) >= last_capture_time:
+                global_log.log(
+                    scalyr_logging.DEBUG_LEVEL_5,
+                    "Performing periodic memory usage capture",
+                )
+                self._capture_snapshot()
+                last_capture_time = int(time.time())
+
+            time.sleep(self._sleep_interval)
+
+    def stop(self):
+        # type: () -> None
+        self._running = False
```

Comment:
Just to capture this... you aren't protected `_running` even though it will be accessed via different threads because `_running` is always updated atomically and there is no requirements of setting multiple variables together in a protected way?